srtcore/api.cpp

```diff
@@ -194,14 +195,8 @@ m_bGCStatus(false),
 m_ClosedSockets()
 {
    // Socket ID MUST start from a random value
-   // Note. Don't use CTimer here, because s_UDTUnited is a static instance of CUDTUnited
-   // with dynamic initialization (calling this constructor), while CTimer has
-   // a static member s_ullCPUFrequency with dynamic initialization.
-   // The order of initialization is not guaranteed.
-   timeval t;
-
-   gettimeofday(&t, 0);
-   srand((unsigned int)t.tv_usec);
+   unsigned seed = count_microseconds(steady_clock::now().time_since_epoch()) & std::numeric_limits<unsigned>::max();
+   srand(seed);
```

Comment:
There may be a problem here.
In POSIX version clock frequency is initialized [here](https://github.com/Haivision/srt/blob/master/srtcore/sync_posix.cpp#L126) in a global context.
It is used by the `count_microseconds` function.

`CUDTUnited` is also initialized in a global context [here](https://github.com/Haivision/srt/blob/master/srtcore/core.cpp#L86).

So if a compiler would decide to construct CUDTUnited before the `sync` module, `s_clock_ticks_per_us` may store some garbage value, and even zero,

Reply:
> There may be a problem here.
> In POSIX version clock frequency is initialized [here](https://github.com/Haivision/srt/blob/master/srtcore/sync_posix.cpp#L126) in a global context.
> It is used by the `count_microseconds` function.
> 
> `CUDTUnited` is also initialized in a global context [here](https://github.com/Haivision/srt/blob/master/srtcore/core.cpp#L86).
> 
> So if a compiler would decide to construct CUDTUnited before the `sync` module, `s_clock_ticks_per_us` may store some garbage value, and even zero,

That's why the commit comes after:

```
Replace global CUDTUnited by static local variable

Static local variable has the benefits of lazy initialization and
avoids the initialization order problem of global variables.
```

`s_clock_ticks_per_us` is initialized before main() or shared library loading, CUDTUnited singleton is initialized at the first time of access. There is no initialization order problem. Now I found another one, the singleton is not thread-safe before C++11:
```
CUDTUnited* CUDT::uglobal() {
    static CUDTUnited instance;
    return &instance;
}
```

pthread_once (or PTHREAD_MUTEX_INITIALIZER) can be used for Unix-like OS. I don't know how to handle that on Windows. Does srt_startup meant to be thread safe?

Reply:
> hat's why the commit comes after:...

Right, I forgot about that change while thinking on `srand`. 🙂 

> Does srt_startup meant to be thread safe?

`srt_startup` is intended to be called only once when an application that uses SRT initializes its resources. E.g. in the beginning on the `main` function.

However, it does not guarantee that no one will try to call it from different threads.
For example, someone might forget to call it, and start creating SRT sockets from different threads. `CUDT::socket()` calls
`uglobal()->startup()` under the hood unless `m_bGCStatus` is available.

So I would say it'd better be thread safe.

Reply:
@quink-black Please note that getting a random integer has been moved to a separate function `srt::sync::genRandomInt(..)` in PR #1997.